tests/FsCodec.SystemTextJson.Tests/UnionConverterTests.fs

```diff
@@ -0,0 +1,443 @@
+
+type FsCheckGenerators =
+    static member CartId = Arb.generate |> Gen.map CartId |> Arb.fromGen
+    static member SkuId = Arb.generate |> Gen.map SkuId |> Arb.fromGen
+
+type DomainPropertyAttribute() =
+    inherit FsCheck.Xunit.PropertyAttribute(QuietOnSuccess = true, Arbitrary=[| typeof<FsCheckGenerators> |])
+
+let roundtripProperty ignoreNulls (profile : JsonSerializerOptions) value =
+    let serialized = JsonSerializer.Serialize(value, profile)
+    render ignoreNulls value =! serialized
+    let deserialized = JsonSerializer.Deserialize<_>(serialized, profile)
+    deserialized =! value
+
+let includeNullsProfile = Options.Create(ignoreNulls = false)
+[<DomainPropertyAttribute(MaxTest=1000)>]
+let ``UnionConverter ignoreNulls Profile roundtrip property test`` (x: TestDU) =
+    let ignoreNulls, profile = false, includeNullsProfile
+    profile.IgnoreNullValues =! false
+    roundtripProperty ignoreNulls profile x
+
+let defaultProfile = Options.Create ()
+[<DomainPropertyAttribute(MaxTest=1000)>]
+let ``UnionConverter opinionated Profile roundtrip property test`` (x: TestDU) =
+    let ignoreNulls, profile = false, defaultProfile
+    profile.IgnoreNullValues =! false
+    roundtripProperty ignoreNulls profile x
+
+//[<Fact>]
+//let ``Implementation ensures no internal errors escape (which would render a WebApi ModelState.Invalid)`` () =
+//    let s = JsonSerializer.CreateDefault()
+//    let mutable gotError = false
+//    s.Error.Add(fun _ -> gotError <- true)
+
+//    let dJson = """{"case":"CaseD","a":"hi"}"""
+//    use dReader = new StringReader(dJson)
+//    use dJsonReader = new JsonTextReader(dReader)
+//    let d = s.Deserialize<TestDU>(dJsonReader)
+
+//    test <@ (CaseD "hi") = d @>
+//    test <@ false = gotError @>
+
+module ``Unmatched case handling`` =
+    [<Fact>]
+    let ``UnionConverter by default throws on unknown cases`` () =
+        let options = Options.Create(UnionConverter ())
+        let aJson = """{"case":"CaseUnknown"}"""
+        let act () = JsonSerializer.Deserialize<TestDU>(aJson, options)
+
+        fun (e : System.InvalidOperationException) -> <@ -1 <> e.Message.IndexOf "No case defined for 'CaseUnknown', and no catchAllCase nominated" @>
+        |> raisesWith <@ act() @>
+
+    [<RequireQualifiedAccess; JsonConverter(typeof<UnionConverter>); JsonUnionConverterOptions("case", CatchAll = "Catchall")>]
+    type DuWithCatchAllWithAttributes =
+    | Known
+    | Catchall
+
+    [<Fact>]
+    let ``UnionConverter supports a nominated catchall via attributes`` () =
+        let aJson = """{"case":"CaseUnknown"}"""
+        let a = JsonSerializer.Deserialize<DuWithCatchAllWithAttributes>(aJson)
+
+        test <@ DuWithCatchAllWithAttributes.Catchall = a @>
+
+    [<RequireQualifiedAccess>]
+    type DuWithCatchAllWithoutAttributes =
+    | Known
+    | Catchall
+
+    [<Fact>]
+    let ``UnionConverter supports a nominated catchall via options`` () =
```

Comment:
not sure it did this before; not sure its a good idea as its a recipe for catchall case not found exceptions

(In general, global converters are a bad idea as they're magic IMO)

Not sure if you agree, but in my book this test should demonstrate random unions don't get affected (people like building SCDU converters etc, but again, I'd prefer not to conflict with such semantics as UnionConverter has plenty semantics just dealing with what it already covers)